### ISSUE TITLE
How to restart my python script running bottle?

### ISSUE THREAD

**Issue by morpheus65535** [bottlepy/bottle#1017]
I want to define a route that can be called to restart my python script.

Here's a sample script:
    from bottle import route, run, redirect
    import os

    @route('/')
    def root():
        return u'test'

    @route('/restart')
    def restart():
        print 'restart'
        os.execlp('python', 'python', __file__)
        redirect ('/')

    run(host='0.0.0.0', port=8080)

I've tried with `os.execlp` but the new instance never reply to request. I just can't access it at all.

Some help would be appreciated to solve this one. Thanks!

**Comment by oz123** [comment]
```
run(host='0.0.0.0', port=8080, reloader=True)
```

**Comment by morpheus65535** [comment]
What is the purpose of this « reloader » argument? How does it help?

**Comment by defnull** [comment]
A process cannot restart itself. You need another process that supervises it and knows how to restart it when necessary.

The bottle 'reloader' does exactly this. If it is enabled, the main process does not start a server but instead starts a child process that then serves requests. If the child process exists with status code 3, it is restarted automatically.

The feature is used to auto-restart the server if any of the imported modules are edited. There is a background-thread in the child process that checks file modification times and calls `sys.exit(3)` if anything changed, causing it to be restarted and loading the new module versions.

There are many other ways do do this. A simple bash script which starts the server in a loop for example, or more sophisticated process supervisors such as supervisord, runit, or systemd.

To sum it up: Stopping a bottle server from within is not a problem with most server backends. Just trigger `sys.exit()`. Restarting it is done from the outside, not the inside.

**Comment by morpheus65535** [comment]
@defnull thanks for the support. Unfortunately, I got no control over the « outside world ». The script coul be run under Windows, Linux, macOS, Docker, etc. I need to stay platform agnostic so a bash script or anything that is not python won’t work.

**Comment by defnull** [comment]
Then use python instead of bash?

**Comment by defnull** [comment]
https://stackoverflow.com/questions/35831291/python-script-to-kill-and-start-itself

**Comment by morpheus65535** [comment]
Seems to be promising. I’m going to try that. Thank you!
